File: runtime/src/protocol.cpp
#include "rivet/protocol.hpp"

#include <algorithm>
#include <array>
#include <climits>
#include <cstring>
#include <stdexcept>
#include <type_traits>
#include <utility>

namespace rivet {
namespace {

constexpr std::array<std::uint8_t, 4> kMagic{'R', 'V', 'T', '1'};

enum class ValueTag : std::uint8_t {
  Null = 0x00,
  False = 0x01,
  True = 0x02,
  Int64 = 0x03,
  String = 0x04,
  Bytes = 0x05,
  List = 0x06,
};

MessageType decode_message_type(std::uint8_t raw) {
  switch (raw) {
    case 1: return MessageType::Hello;
    case 2: return MessageType::Request;
    case 3: return MessageType::Response;
    case 4: return MessageType::Error;
    case 5: return MessageType::Event;
    case 6: return MessageType::Cancel;
    case 7: return MessageType::Shutdown;
    default: throw std::runtime_error("unknown Rivet message type");
  }
}

bool is_continuation(std::uint8_t byte) {
  return (byte & 0xc0u) == 0x80u;
}

bool valid_utf8(std::uint8_t const* data, std::size_t size) {
  std::size_t i = 0;
  while (i < size) {
    auto const first = data[i];
    if (first <= 0x7f) {
      ++i;
      continue;
    }
    if (first >= 0xc2 && first <= 0xdf) {
      if (i + 1 >= size || !is_continuation(data[i + 1])) return false;
      i += 2;
      continue;
    }
    if (first == 0xe0) {
      if (i + 2 >= size || data[i + 1] < 0xa0 || data[i + 1] > 0xbf ||
          !is_continuation(data[i + 2])) return false;
      i += 3;
      continue;
    }
    if ((first >= 0xe1 && first <= 0xec) || (first >= 0xee && first <= 0xef)) {
      if (i + 2 >= size || !is_continuation(data[i + 1]) ||
          !is_continuation(data[i + 2])) return false;
      i += 3;
      continue;
    }
    if (first == 0xed) {
      if (i + 2 >= size || data[i + 1] < 0x80 || data[i + 1] > 0x9f ||
          !is_continuation(data[i + 2])) return false;
      i += 3;
      continue;
    }
    if (first == 0xf0) {
      if (i + 3 >= size || data[i + 1] < 0x90 || data[i + 1] > 0xbf ||
          !is_continuation(data[i + 2]) || !is_continuation(data[i + 3])) return false;
      i += 4;
      continue;
    }
    if (first >= 0xf1 && first <= 0xf3) {
      if (i + 3 >= size || !is_continuation(data[i + 1]) ||
          !is_continuation(data[i + 2]) || !is_continuation(data[i + 3])) return false;
      i += 4;
      continue;
    }
    if (first == 0xf4) {
      if (i + 3 >= size || data[i + 1] < 0x80 || data[i + 1] > 0x8f ||
          !is_continuation(data[i + 2]) || !is_continuation(data[i + 3])) return false;
      i += 4;
      continue;
    }
    return false;
  }
  return true;
}

void append_u32(Bytes& out, std::uint32_t value) {
  for (int i = 0; i < 4; ++i) {
    out.push_back(static_cast<std::uint8_t>((value >> (8 * i)) & 0xff));
  }
}

void append_u64(Bytes& out, std::uint64_t value) {
  for (int i = 0; i < 8; ++i) {
    out.push_back(static_cast<std::uint8_t>((value >> (8 * i)) & 0xff));
  }
}

std::uint32_t read_u32(std::uint8_t const* data) {
  std::uint32_t value = 0;
  for (int i = 0; i < 4; ++i) {
    value |= static_cast<std::uint32_t>(data[i]) << (8 * i);
  }
  return value;
}

std::uint64_t read_u64(std::uint8_t const* data) {
  std::uint64_t value = 0;
  for (int i = 0; i < 8; ++i) {
    value |= static_cast<std::uint64_t>(data[i]) << (8 * i);
  }
  return value;
}

class Reader {
 public:
  explicit Reader(Bytes const& bytes) : bytes_(bytes) {}

  std::uint8_t byte() {
    require(1);
    return bytes_[offset_++];
  }

  std::uint32_t u32() {
    require(4);
    auto const value = read_u32(bytes_.data() + offset_);
    offset_ += 4;
    return value;
  }

  std::uint64_t u64() {
    require(8);
    auto const value = read_u64(bytes_.data() + offset_);
    offset_ += 8;
    return value;
  }

  Bytes bytes(std::size_t size) {
    require(size);
    Bytes result(bytes_.begin() + static_cast<std::ptrdiff_t>(offset_),
                 bytes_.begin() + static_cast<std::ptrdiff_t>(offset_ + size));
    offset_ += size;
    return result;
  }

  bool empty() const noexcept { return offset_ == bytes_.size(); }
  std::size_t remaining() const noexcept { return bytes_.size() - offset_; }

 private:
  void require(std::size_t size) const {
    if (size > bytes_.size() - offset_) {
      throw std::runtime_error("truncated Rivet value payload");
    }
  }

  Bytes const& bytes_;
  std::size_t offset_{};
};

void consume_encode_node(std::size_t& remaining_nodes) {
  if (remaining_nodes == 0) {
    throw std::length_error("Rivet value node count exceeds protocol limit");
  }
  --remaining_nodes;
}

void consume_decode_node(std::size_t& remaining_nodes) {
  if (remaining_nodes == 0) {
    throw std::runtime_error("Rivet value node count exceeds protocol limit");
  }
  --remaining_nodes;
}

void encode_into(Bytes& out,
                 Value const& value,
                 std::size_t depth,
                 std::size_t& remaining_nodes) {
  consume_encode_node(remaining_nodes);
  std::visit(
      [&](auto const& data) {
        using T = std::decay_t<decltype(data)>;
        if constexpr (std::is_same_v<T, std::monostate>) {
          out.push_back(static_cast<std::uint8_t>(ValueTag::Null));
        } else if constexpr (std::is_same_v<T, bool>) {
          out.push_back(static_cast<std::uint8_t>(data ? ValueTag::True : ValueTag::False));
        } else if constexpr (std::is_same_v<T, std::int64_t>) {
          out.push_back(static_cast<std::uint8_t>(ValueTag::Int64));
          std::uint64_t bits = 0;
          static_assert(sizeof(bits) == sizeof(data));
          std::memcpy(&bits, &data, sizeof(bits));
          append_u64(out, bits);
        } else if constexpr (std::is_same_v<T, std::string>) {
          auto const* raw = reinterpret_cast<std::uint8_t const*>(data.data());
          if (!valid_utf8(raw, data.size())) {
            throw std::runtime_error("invalid UTF-8 in Rivet string");
          }
          out.push_back(static_cast<std::uint8_t>(ValueTag::String));
          if (data.size() > UINT32_MAX) {
            throw std::length_error("Rivet string exceeds protocol limit");
          }
          append_u32(out, static_cast<std::uint32_t>(data.size()));
          out.insert(out.end(), data.begin(), data.end());
        } else if constexpr (std::is_same_v<T, Bytes>) {
          out.push_back(static_cast<std::uint8_t>(ValueTag::Bytes));
          if (data.size() > UINT32_MAX) {
            throw std::length_error("Rivet byte vector exceeds protocol limit");
          }
          append_u32(out, static_cast<std::uint32_t>(data.size()));
          out.insert(out.end(), data.begin(), data.end());
        } else if constexpr (std::is_same_v<T, Value::List>) {
          if (depth >= kMaxValueDepth) {
            throw std::length_error("Rivet value nesting exceeds protocol limit");
          }
          if (data.size() > UINT32_MAX) {
            throw std::length_error("Rivet list exceeds protocol limit");
          }
          if (data.size() > remaining_nodes) {
            throw std::length_error("Rivet value node count exceeds protocol limit");
          }
          out.push_back(static_cast<std::uint8_t>(ValueTag::List));
          append_u32(out, static_cast<std::uint32_t>(data.size()));
          for (auto const& item : data) {
            encode_into(out, item, depth + 1, remaining_nodes);
          }
        }
      },
      value.data);
}

Value decode_one(Reader& reader,
                 std::size_t depth,
                 std::size_t& remaining_nodes) {
  consume_decode_node(remaining_nodes);
  auto const tag = static_cast<ValueTag>(reader.byte());
  switch (tag) {
    case ValueTag::Null:
      return Value{};
    case ValueTag::False:
      return Value(false);
    case ValueTag::True:
      return Value(true);
    case ValueTag::Int64: {
      auto const bits = reader.u64();
      std::int64_t value = 0;
      static_assert(sizeof(bits) == sizeof(value));
      std::memcpy(&value, &bits, sizeof(value));
      return Value(value);
    }
    case ValueTag::String: {
      auto const raw = reader.bytes(reader.u32());
      if (!valid_utf8(raw.data(), raw.size())) {
        throw std::runtime_error("invalid UTF-8 in Rivet string");
      }
      return Value(std::string(raw.begin(), raw.end()));
    }
    case ValueTag::Bytes:
      return Value(reader.bytes(reader.u32()));
    case ValueTag::List: {
      if (depth >= kMaxValueDepth) {
        throw std::runtime_error("Rivet value nesting exceeds protocol limit");
      }
      Value::List list;
      auto const count = reader.u32();
      // Every declared element consumes at least one value node. Enforce this
      // before reserving memory so a tiny payload cannot request a huge vector.
      if (count > remaining_nodes) {
        throw std::runtime_error("Rivet value node count exceeds protocol limit");
      }
      // Every encoded item also needs at least one tag byte.
      if (count > reader.remaining()) {
        throw std::runtime_error("impossible Rivet list length");
      }
      list.reserve(count);
      for (std::uint32_t i = 0; i < count; ++i) {
        list.push_back(decode_one(reader, depth + 1, remaining_nodes));
      }
      return Value(std::move(list));
    }
  }
  throw std::runtime_error("unknown Rivet value tag");
}

}  // namespace

void write_frame(Transport& transport, Frame const& frame) {
  (void)decode_message_type(static_cast<std::uint8_t>(frame.type));
  if (frame.payload.size() > kMaxFramePayloadSize) {
    throw std::length_error("Rivet frame payload exceeds protocol limit");
  }

  Bytes header;
  header.reserve(18);
  header.insert(header.end(), kMagic.begin(), kMagic.end());
  header.push_back(kProtocolVersion);
  header.push_back(static_cast<std::uint8_t>(frame.type));
  append_u64(header, frame.id);
  append_u32(header, static_cast<std::uint32_t>(frame.payload.size()));

  transport.write_all(header.data(), header.size());
  if (!frame.payload.empty()) {
    transport.write_all(frame.payload.data(), frame.payload.size());
  }
  transport.flush();
}

std::optional<Frame> read_frame(Transport& transport) {
  std::array<std::uint8_t, 18> header{};
  if (!transport.read_exact(header.data(), header.size())) {
    return std::nullopt;
  }

  if (!std::equal(kMagic.begin(), kMagic.end(), header.begin())) {
    throw std::runtime_error("invalid Rivet frame magic");
  }
  if (header[4] != kProtocolVersion) {
    throw std::runtime_error("unsupported Rivet protocol version");
  }

  Frame frame;
  frame.type = decode_message_type(header[5]);
  frame.id = read_u64(header.data() + 6);
  auto const payload_size = read_u32(header.data() + 14);
  if (payload_size > kMaxFramePayloadSize) {
    throw std::length_error("Rivet frame payload exceeds protocol limit");
  }
  frame.payload.resize(payload_size);
  if (payload_size != 0 &&
      !transport.read_exact(frame.payload.data(), frame.payload.size())) {
    throw std::runtime_error("unexpected EOF in Rivet frame payload");
  }
  return frame;
}

Bytes encode_value(Value const& value) {
  Bytes result;
  std::size_t remaining_nodes = kMaxValueNodes;
  encode_into(result, value, 0, remaining_nodes);
  return result;
}

Value decode_value(Bytes const& bytes) {
  Reader reader(bytes);
  std::size_t remaining_nodes = kMaxValueNodes;
  auto value = decode_one(reader, 0, remaining_nodes);
  if (!reader.empty()) {
    throw std::runtime_error("trailing bytes after Rivet value");
  }
  return value;
}

}  // namespace rivet

File: platform/macos/Sources/RivetRuntime/Protocol.swift
import Foundation

public let rivetProtocolVersion: UInt8 = 1
public let rivetMaxFramePayloadSize = 64 * 1024 * 1024
public let rivetMaxValueDepth = 64
public let rivetMaxValueNodes = 1 << 18

public enum RivetMessageType: UInt8, Sendable {
    case hello = 1
    case request = 2
    case response = 3
    case error = 4
    case event = 5
    case cancel = 6
    case shutdown = 7
}

public enum RivetValue: Equatable, Sendable {
    case null
    case bool(Bool)
    case int64(Int64)
    case string(String)
    case bytes(Data)
    case list([RivetValue])
}

public struct RivetFrame: Equatable, Sendable {
    public var type: RivetMessageType
    public var id: UInt64
    public var payload: Data

    public init(type: RivetMessageType, id: UInt64, payload: Data = Data()) {
        self.type = type
        self.id = id
        self.payload = payload
    }
}

public enum RivetProtocolError: Error, Equatable, CustomStringConvertible {
    case truncated
    case invalidMagic
    case unsupportedVersion(UInt8)
    case unknownMessageType(UInt8)
    case unknownValueTag(UInt8)
    case invalidUTF8
    case trailingBytes
    case lengthOverflow
    case nestingTooDeep

    public var description: String {
        switch self {
        case .truncated: return "truncated Rivet payload"
        case .invalidMagic: return "invalid Rivet frame magic"
        case .unsupportedVersion(let version): return "unsupported Rivet protocol version \(version)"
        case .unknownMessageType(let value): return "unknown Rivet message type \(value)"
        case .unknownValueTag(let value): return "unknown Rivet value tag \(value)"
        case .invalidUTF8: return "invalid UTF-8 in Rivet string"
        case .trailingBytes: return "trailing bytes after Rivet value"
        case .lengthOverflow: return "Rivet value exceeds protocol length limit"
        case .nestingTooDeep: return "Rivet value nesting exceeds protocol limit"
        }
    }
}

private enum ValueTag: UInt8 {
    case null = 0x00
    case falseValue = 0x01
    case trueValue = 0x02
    case int64 = 0x03
    case string = 0x04
    case bytes = 0x05
    case list = 0x06
}

private let magic = Data([0x52, 0x56, 0x54, 0x31]) // RVT1

private extension Data {
    mutating func appendLE<T: FixedWidthInteger>(_ value: T) {
        var little = value.littleEndian
        Swift.withUnsafeBytes(of: &little) { append(contentsOf: $0) }
    }
}

private struct Reader {
    let data: Data
    var offset: Int = 0

    mutating func readByte() throws -> UInt8 {
        guard offset < data.count else { throw RivetProtocolError.truncated }
        defer { offset += 1 }
        return data[offset]
    }

    mutating func readInteger<T: FixedWidthInteger>(_ type: T.Type) throws -> T {
        let width = MemoryLayout<T>.size
        guard offset <= data.count - width else { throw RivetProtocolError.truncated }
        var result: T = 0
        for i in 0..<width {
            result |= T(data[offset + i]) << T(i * 8)
        }
        offset += width
        return result
    }

    mutating func readData(count: Int) throws -> Data {
        guard count >= 0, offset <= data.count - count else {
            throw RivetProtocolError.truncated
        }
        let result = data.subdata(in: offset..<(offset + count))
        offset += count
        return result
    }

    var isAtEnd: Bool { offset == data.count }
    var remaining: Int { data.count - offset }
}

private func ensureEncodeCapacity(_ count: Int, output: Data) throws {
    guard count >= 0,
          output.count <= rivetMaxFramePayloadSize,
          count <= rivetMaxFramePayloadSize - output.count else {
        throw RivetProtocolError.lengthOverflow
    }
}

public func encodeRivetValue(_ value: RivetValue) throws -> Data {
    var result = Data()
    var remainingNodes = rivetMaxValueNodes
    try encode(value, into: &result, depth: 0, remainingNodes: &remainingNodes)
    return result
}

private func encode(
    _ value: RivetValue,
    into output: inout Data,
    depth: Int,
    remainingNodes: inout Int
) throws {
    guard remainingNodes > 0 else { throw RivetProtocolError.lengthOverflow }
    remainingNodes -= 1

    switch value {
    case .null:
        try ensureEncodeCapacity(1, output: output)
        output.append(ValueTag.null.rawValue)
    case .bool(false):
        try ensureEncodeCapacity(1, output: output)
        output.append(ValueTag.falseValue.rawValue)
    case .bool(true):
        try ensureEncodeCapacity(1, output: output)
        output.append(ValueTag.trueValue.rawValue)
    case .int64(let value):
        try ensureEncodeCapacity(9, output: output)
        output.append(ValueTag.int64.rawValue)
        output.appendLE(UInt64(bitPattern: value))
    case .string(let value):
        let byteCount = value.utf8.count
        guard byteCount <= Int(UInt32.max) else { throw RivetProtocolError.lengthOverflow }
        try ensureEncodeCapacity(5 + byteCount, output: output)
        output.append(ValueTag.string.rawValue)
        output.appendLE(UInt32(byteCount))
        output.append(contentsOf: value.utf8)
    case .bytes(let bytes):
        guard bytes.count <= Int(UInt32.max) else { throw RivetProtocolError.lengthOverflow }
        try ensureEncodeCapacity(5 + bytes.count, output: output)
        output.append(ValueTag.bytes.rawValue)
        output.appendLE(UInt32(bytes.count))
        output.append(bytes)
    case .list(let values):
        guard depth < rivetMaxValueDepth else { throw RivetProtocolError.nestingTooDeep }
        guard values.count <= Int(UInt32.max) else { throw RivetProtocolError.lengthOverflow }
        guard values.count <= remainingNodes else { throw RivetProtocolError.lengthOverflow }
        try ensureEncodeCapacity(5, output: output)
        output.append(ValueTag.list.rawValue)
        output.appendLE(UInt32(values.count))
        for value in values {
            try encode(value, into: &output, depth: depth + 1, remainingNodes: &remainingNodes)
        }
    }
}

public func decodeRivetValue(_ data: Data) throws -> RivetValue {
    guard data.count <= rivetMaxFramePayloadSize else {
        throw RivetProtocolError.lengthOverflow
    }
    var reader = Reader(data: data)
    var remainingNodes = rivetMaxValueNodes
    let value = try decodeValue(from: &reader, depth: 0, remainingNodes: &remainingNodes)
    guard reader.isAtEnd else { throw RivetProtocolError.trailingBytes }
    return value
}

private func decodeValue(
    from reader: inout Reader,
    depth: Int,
    remainingNodes: inout Int
) throws -> RivetValue {
    guard remainingNodes > 0 else { throw RivetProtocolError.lengthOverflow }
    remainingNodes -= 1

    let rawTag = try reader.readByte()
    guard let tag = ValueTag(rawValue: rawTag) else {
        throw RivetProtocolError.unknownValueTag(rawTag)
    }

    switch tag {
    case .null:
        return .null
    case .falseValue:
        return .bool(false)
    case .trueValue:
        return .bool(true)
    case .int64:
        let bits: UInt64 = try reader.readInteger(UInt64.self)
        return .int64(Int64(bitPattern: bits))
    case .string:
        let length: UInt32 = try reader.readInteger(UInt32.self)
        let bytes = try reader.readData(count: Int(length))
        guard let value = String(data: bytes, encoding: .utf8) else {
            throw RivetProtocolError.invalidUTF8
        }
        return .string(value)
    case .bytes:
        let length: UInt32 = try reader.readInteger(UInt32.self)
        return .bytes(try reader.readData(count: Int(length)))
    case .list:
        guard depth < rivetMaxValueDepth else { throw RivetProtocolError.nestingTooDeep }
        let count: UInt32 = try reader.readInteger(UInt32.self)
        guard Int(count) <= remainingNodes else { throw RivetProtocolError.lengthOverflow }
        guard Int(count) <= reader.remaining else {
            throw RivetProtocolError.truncated
        }
        var values: [RivetValue] = []
        values.reserveCapacity(Int(count))
        for _ in 0..<count {
            values.append(try decodeValue(
                from: &reader,
                depth: depth + 1,
                remainingNodes: &remainingNodes
            ))
        }
        return .list(values)
    }
}

public func encodeRivetFrame(_ frame: RivetFrame) throws -> Data {
    guard frame.payload.count <= rivetMaxFramePayloadSize else {
        throw RivetProtocolError.lengthOverflow
    }

    var result = Data()
    result.append(magic)
    result.append(rivetProtocolVersion)
    result.append(frame.type.rawValue)
    result.appendLE(frame.id)
    result.appendLE(UInt32(frame.payload.count))
    result.append(frame.payload)
    return result
}

public func decodeRivetFrame(_ data: Data) throws -> RivetFrame {
    var reader = Reader(data: data)
    guard try reader.readData(count: 4) == magic else {
        throw RivetProtocolError.invalidMagic
    }

    let version = try reader.readByte()
    guard version == rivetProtocolVersion else {
        throw RivetProtocolError.unsupportedVersion(version)
    }

    let rawType = try reader.readByte()
    guard let type = RivetMessageType(rawValue: rawType) else {
        throw RivetProtocolError.unknownMessageType(rawType)
    }

    let id: UInt64 = try reader.readInteger(UInt64.self)
    let length: UInt32 = try reader.readInteger(UInt32.self)
    guard Int(length) <= rivetMaxFramePayloadSize else {
        throw RivetProtocolError.lengthOverflow
    }
    let payload = try reader.readData(count: Int(length))
    guard reader.isAtEnd else { throw RivetProtocolError.trailingBytes }
    return RivetFrame(type: type, id: id, payload: payload)
}
